---
title: "5hmC"
author: "Rachel Edgar"
date: "02/05/2017"
output: html_document
---

```{r setup, include=FALSE}
library(methylumi)
library(wateRmelon)
library(RPMM)
library("ggsci")
library(RColorBrewer)


setwd("~/Side/5hmC_Turecki/")
```

## Load the data from Genome Studio and make methylumi objects

```{r,eval=F, echo=T}
allFile <- ("/home/redgar/Side/5hmC_Turecki/5hmC pilot with Turecki samples/genomestudio files/5hmC pilot-alldata.txt") 
qcFile <-  ("/home/redgar/Side/5hmC_Turecki/5hmC pilot with Turecki samples/genomestudio files/5hmC pilot-qcfile.txt")

#Producing the lumi objects
hmc<- lumiMethyR(allFile)
hmc.2 <- methylumiR(allFile, qcfile = qcFile)
#save(hmc,hmc.2, file="hmc_methlylumi.RData")
```


```{r, echo=T}
load("~/Side/5hmC_Turecki/hmc_methlylumi.RData")
```


## Meta data check samples match
```{r}
meta_hmc<-read.table("/home/redgar/Side/5hmC_Turecki/5hmC pilot with Turecki samples/genomestudio files/5hmC pilot-samplefile.txt", sep="\t", header=T)
meta_hmc$Subject<-c("64","104","104","173","169","173","64","169")
meta_hmc$DNAm_type<-c("5mc","5mc","5mc+5hmc","5mc","5mc","5mc+5hmc","5mc+5hmc","5mc+5hmc")

identical(sampleNames(hmc.2), as.character(meta_hmc$Sample_ID))
```

#### Pvalues of samples on each array
```{r}
avgPval <- colMeans(pvals(hmc.2))

#p-values for detection for the samples
meta_hmc$Det_pval<-avgPval
ggplot(meta_hmc)+  geom_point(aes(as.factor(Sample_ID), Det_pval, fill=Sample_ID), shape=21, size=3)+theme_bw()
```


#### Beta distribution
```{r}
hmcBetas<-betas(hmc.2)
Beta_sample<-hmcBetas[sample(1:nrow(hmcBetas), 100000),]

## Beta distribtuions 
Beta_sample_melted<- melt(Beta_sample)
#remove NAs before plotting (otherwise get many non-inifnite warnings)
Beta_Plot<-Beta_sample_melted[which(!(is.na(Beta_sample_melted$value))),]

#add meta
Beta_Plot<-merge(Beta_Plot,meta_hmc, by.x="Var2", by.y="Sample_ID")

ggplot(Beta_Plot, aes(value, group=Var2, color=DNAm_type))+
  geom_density(size=1)+theme_bw()+scale_color_manual(values=c("red","grey"))

```


#### Confirm ID with SNPs and cluster by DNAm type?
```{r}
# Clustering By Any Meta Data variable
# remove rows with NAs
hmcbeta_cluster<-hmcBetas[complete.cases(hmcBetas),]

# plot clustering with color function
plotHclustColors <- function(matrix,leafcolor) {
  colnames(matrix) <- leafcolor
  d <- dist(t(matrix))
  hc <- hclust(d, method = "complete") #single, complete, average, ward
  color<-rep(brewer.pal(12,"Paired"),ceiling(length(unique(leafcolor))/12))
  labelColors <- color[sample(1:length(color),length(unique(leafcolor)))]
  colLab <- function(n) {
    if (is.leaf(n)) {
      a <- attributes(n)
      labCol <- labelColors[which(unique(leafcolor) == a$label)]
      attr(n, "nodePar") <- c(a$nodePar, lab.col=labCol)
    }
    n
  }
  clusDendro <- dendrapply(as.dendrogram(hc), colLab)
  plot(clusDendro)
}


# Plot
plotHclustColors(hmcbeta_cluster, meta_hmc$DNAm_type) 


#Genotyping Probes (show that samples from the same individual cluster by SNPs)
CpGs<- rownames(hmcBetas)
SNP_Probes<-CpGs[grep("rs", CpGs)]# 59 SNP probes on EPIC
SNPs<-hmcBetas[CpGs%in%SNP_Probes,]
# Remove NAs (if any)
SNPs<-SNPs[complete.cases(SNPs),]# 58 cause one was all NA

plotHclustColors(SNPs, meta_hmc$Subject)#all samples pair with their other time point
```


# Probe Filtering
##### Removal of SNP Probes
We remove the SNP probes as they are used as an internal control to ensure your samples are what you think they are and are not used for any methylation analysis.

```{r, echo=FALSE}
hmc.2_filtered <- hmc.2[substring(featureNames(hmc.2),1,2) != "rs", ]
dim(hmc.2_filtered) # probes = 865859, n = 8
```


##### Cross-hybridizing probes and polymorphic probes. Don't have an a Magda Price annotation for EPIC. So gonna go find that shit. 
https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1066-1
"43,254 cross-reactive probes with â‰¥ 47 bp homology with an off-target site, of which 15,782 (36.5 %) are new to the EPIC platform"

They include this annotated list in their supplement.

```{r}
cross_reactive<-read.csv("~/Side/5hmC_Turecki/Pidsley_cross_reactive.csv", stringsAsFactors = F)
hmc.2_filtered<-hmc.2_filtered[which(!(featureNames(hmc.2_filtered)%in%cross_reactive$X)),]
dim(hmc.2_filtered) # probes = 822682, n = 8, 43177 filtered
```

For polymorphic probes I will The Pidsley annotation aswell for "Probes overlapping genetic variants at targeted CpG sites." and "Probes overlapping genetic variants at single base extension sites for Infinium Type I probes" but NOT "Probes with genetic variants overlapping the body of the probe: 48 base pairs for Infinium Type I probes and 49 base pairs for Infinium Type II probes."

```{r}
polymorphic<-read.csv("~/Side/5hmC_Turecki/Pidsley_Polymorphic_CpGs.csv",stringsAsFactors = F)
length(unique(polymorphic$PROBE))
baseext<-read.csv("~/Side/5hmC_Turecki/Pidsley_single_base_extension.csv",stringsAsFactors = F)
length(unique(baseext$PROBE))

hmc.2_filtered<-hmc.2_filtered[which(!(featureNames(hmc.2_filtered)%in%c(polymorphic$PROBE, baseext$PROBE))),]
dim(hmc.2_filtered) # probes = 811063, n = 8, 11619 filtered
```

##### Sex Chromosomes
All males were run and 5hmC shows cool trends at X chromsomes so I am recklessly leaving them in!
```{r}
table(featureData(hmc.2_filtered)$CHR)

      ## But if you wanted to 
      # hmc.2_filtered <- hmc.2_filtered[!featureData(hmc.2_filtered)$CHR%in%c("X", "Y"), ]
      # dim(hmc.2_filtered) 
```


We have removed 54,855 probes. This leaves us with 811,063 probes for our analysis. Including 17,745 X chr and 379 Y chr CpGs.



# Watermelon Filter and Normalization
# BMIQ (probe type normalization)
    #Checks
      #iDMR, X, Snps
      #Replicates correlation

USing pfilter from wateRmelon to filter this
```{r}
#perc remove samples having this percentage of sites with a detection p-value greater than pnthresh, default set to 1 (pnthresh=0.05), gonna to use 5% here with hmC data because it has extra bad detection values but genrally stick with default.
hmc.pf<-pfilter(hmc.2_filtered, perc=5)

    # 0 samples having 5 % of sites with a detection p-value greater than 0.05 were removed 
    # Samples removed:  
    # 17663 sites were removed as beadcount <3 in 5 % of samples 
    # 26279 sites having 1 % of samples with a detection p-value greater than 0.05 were removed 
dim(hmc.pf) #  770,004      8 
# the adjusted filter changes the CpG number filtered as well becuase the 3 baddies change the pvalue counts for probes.

hmc.2_filtered<-hmc.pf
#save(hmc.2_filtered, file = "hmC_fully_filtered.RData")
```

We have removed another 41,059 probes. This leaves us with 770,004 probes for our analysis.  Including 16,106 X chr and 363 Y chr CpGs.



#### Probe attrition plot
```{r}
df<-data.frame(sample_num_remaining=c(865918,865859,822682,811063,793400,770004), filter=c("EPIC Probe Number","Removal of SNP Probes","Removal of Pidsley Cross Reactive Probes","Removal of Pidsley Polymorphic Probes", "Removal of Probes with Beadcount <3\nin 5 % of Samples", "Removal of Probes with 1 % of samples\nwith a detection p-value greater than 0.05"))
df$sample_num_lost<-c(0,sapply(2:nrow(df), function(x) df$sample_num_remaining[x-1]-df$sample_num_remaining[x]))

df$filter<-factor(df$filter, rev(df$filter))
          
library(scales)
ggplot(df)+
  geom_bar(aes(filter,-sample_num_remaining), stat="identity", fill="grey70", color="black")+
  geom_bar(aes(filter,sample_num_lost), stat="identity",fill="darkred", color="black")+
  geom_text(aes(x=filter, y=-min(sample_num_remaining)/2,  label=comma(sample_num_remaining)))+
  geom_text(aes(x=filter, y=max(sample_num_lost)/1.5,  label=comma(sample_num_lost)))+
  geom_hline(yintercept=0)+
  coord_flip()+theme_bw()+ylab("")+xlab("")+
  theme(axis.line = element_blank(),
        axis.ticks = element_blank(),
        axis.text.x = element_blank(),
        axis.text.y = element_text(colour = "grey20", size=12),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank()) +
  scale_x_discrete(position = "top") 
```






